### PR TITLE
[SSP] Model (scheduling) operations and dependences.

### PR DESCRIPTION
**Pull request by jopperm**
This PR introduces the `OperationOp`, which represents a vertex and its incoming edges in the dependence graph.

### CODE REVIEW

File: lib/Dialect/SSP/SSPOps.cpp [hunk 1]
```diff
@@ -49,6 +49,255 @@ DependenceGraphOp InstanceOp::getDependenceGraph() {
...
+  auto parsePropertiesResult =
+      parseOptionalPropertyArray(properties, parser, alreadyParsed);
+  if (parsePropertiesResult.hasValue()) {
+    if (failed(*parsePropertiesResult))
+      return failure();
+    result.addAttribute(builder.getStringAttr("properties"), properties);
+  }
+
+  // Parse default attr-dict
+  (void)parser.parseOptionalAttrDict(result.attributes);
```
Comment (mortbopet) [inline lib/Dialect/SSP/SSPOps.cpp:133]: Is this the correct way of handling `parseOptionalAttrDict`? It doesn't return a `OptionalParseResult` so I'm assuming if `parseOptionalAttrDict` returns failure, something actually failed.

File: lib/Dialect/SSP/SSPOps.cpp [hunk 2]
```diff
@@ -49,6 +49,255 @@ DependenceGraphOp InstanceOp::getDependenceGraph() {
...
+  int lastIdx = -1;
+  for (auto dep : dependences.getAsRange<DependenceAttr>()) {
+    int idx = dep.getOperandIdx();
+    FlatSymbolRefAttr sourceRef = dep.getSourceRef();
+
+    if (!sourceRef) {
+      // Def-use deps use the index to refer to one of the SSA operands.
+      if (idx >= nOperands)
+        return emitError(
+            "Operand index is out of bounds for def-use dependence attribute");
```
Comment (mortbopet) [inline lib/Dialect/SSP/SSPOps.cpp:237]: Missing test

File: lib/Dialect/SSP/SSPOps.cpp [hunk 3]
```diff
@@ -49,6 +49,255 @@ DependenceGraphOp InstanceOp::getDependenceGraph() {
...
+
+      // Indices may be sparse, but shall be sorted and unique.
+      if (idx <= lastIdx)
+        return emitError("Def-use operand indices in dependence attribute are "
```
Comment (mortbopet) [inline lib/Dialect/SSP/SSPOps.cpp:241]: missing test

File: lib/Dialect/SSP/SSPOps.cpp [hunk 4]
```diff
@@ -49,6 +49,255 @@ DependenceGraphOp InstanceOp::getDependenceGraph() {
...
+    } else {
+      // Auxiliary deps are expected to follow the def-use deps (if present),
+      // and hence use indices >= #operands.
+      if (idx < nOperands)
+        return emitError() << "Auxiliary dependence from " << sourceRef
+                           << " is interleaved with SSA operands";
+
+      // Indices shall be consecutive (special case: the first aux dep)
+      if (!((idx == lastIdx + 1) || (idx > lastIdx && idx == nOperands)))
+        return emitError("Auxiliary operand indices in dependence attribute "
```
Comment (mortbopet) [inline lib/Dialect/SSP/SSPOps.cpp:252]: Missing test

File: lib/Dialect/SSP/SSPOps.cpp [hunk 5]
```diff
@@ -49,6 +49,255 @@ DependenceGraphOp InstanceOp::getDependenceGraph() {
...
+        return emitError(
+                   "Auxiliary dependence references invalid source operation: ")
+               << sourceRef;
+      }
+    }
+
+  // If a linkedOperatorType property is present, verify that it references a
+  // valid operator type.
+  if (ArrayAttr properties = getPropertiesAttr())
+    for (auto prop : properties.getAsRange<Attribute>())
```
Comment (mortbopet) [inline lib/Dialect/SSP/SSPOps.cpp:286]: I'd prefer brackets for the `if` and `for` here.

**Review by mortbopet — APPROVED** [review]
Those are some beasts of a parser and printer, but I see the necessity due to the generality of the operation. A few nits, but LGTM!

File: lib/Dialect/SSP/SSPOps.cpp (hunk 1 above)
Comment (jopperm) [reply lib/Dialect/SSP/SSPOps.cpp]: You're right, thanks!

**Pull request merged**